### PR TITLE
Add unsafeStatic values

### LINKED ISSUE #78: Allow injection of static content into templates
As discussed in [this thread](https://twitter.com/BedeOverend/status/902308389539086336), it would be useful to have some way of injecting static content into the templates that would be used at compile time of the templates, but not change per-render.

Potential syntax that come out of that thread would be using a utility function to demarcate static content:

```js
import { html, unsafeStatic } from 'lit-html'
import SomeElement from 'some-element';

const elementName = unsafeStatic(SomeElement.is);
let dynamicTagTemplate = html`<${elementName}>Hello</${elementName}>`
```

/cc @justinfagnani let me know if I misinterpreted that syntax you suggested

### LINKED ISSUE #218: Sharing styles
One of the benefits of a js templating approach is that it is easier to share code between js, css and html. I work a lot of css animations and I want to, for example, share animation timing constants between my css and js. However, the following does not work currently within lit:

```
const timing = 500;
const template = html`
  <style>
    .some-element {
      ...some styling
      animation-duration: ${timing}ms;
    }
  </style>
`;

setTimeout(() => {
   ...do some work after animation
}, timing);
```

As the ${...} gets parsed by lit. This behavior is good for html, as we want to prevent XSS. But does the same apply for styles?

### PR DESCRIPTION
**Pull request by justinfagnani**
Fixes #78 
Fixes #218

### CODE REVIEW

**Comment by daKmoR** [comment]
that is looking really nice :) - perfect timing :)
will try it out next thing in the morning :)

File: test/index.html [hunk 1]
```diff
@@ -14,9 +14,9 @@
     <script type="module" src="./lib/async-append_test.js"></script>
     <script type="module" src="./lib/async-replace_test.js"></script>
     <script>
-      WCT.loadSuites([
-        'shady.html',
-      ]);
+      // WCT.loadSuites([
```
Comment (stramel) [inline test/index.html]: Did you mean to comment out the tests?

**Comment by Lodin** [comment]
Any updates on this pull request?

**Comment by sorvell** [comment]
I've done a bit of ad hoc perf testing on this and it looked roughly like a 5-10% hit to initial rendering performance even without using the feature. This seems pretty unfortunate. 

The large majority of this cost looked to be in key generation in the `TemplateResult` constructor. I focused on initial render but I expect also to see perhaps a similar penalty for update since `TemplateResult`s are created then as well.

Some ideas to consider to improve this:
1. The key generation could be made faster. It didn't look to me like it could be improved very much, but this may be worth more investigation.

1. Make the cost more pay-for-play by making the `TemplateResult` key generate using the slow path (iterating and building in statics) only if `unsafeStatic` has been used and use the previous method otherwise (just the strings array passed to the tag function). It seems like this would need some kind of heuristic, something like: go into slow mode when `unsafeStatic` is called and go back to fast mode after the next `TemplateResult` key has been generated. Alternatively, pay-for-play could more explicitly be done via a new tag function, say,  `htmlWithStatics`. You'd have to use this to get `unsafeStatic` support and this would trigger the slower key generation. Introducing another base level API seems fairly unfortunate, however.

1. Or we could allow the user to supply their on result key. If the user supplies a key, the key generation would be skipped entirely. This is fairly ugly since it mars the API, but it would be fast. Note, the key generation would have to be moved out of the `TemplateResult` constructor to accomplish this. Imagine a `keyedResult` function that would associate the result with the given key. The unfortunate thing here is you'd need a key with every result to get a speed up (or maybe you could only use statics if you supply a key), but something like: 

```keyedResult(html`<div>${unsafeStatic('foo')}</div>`, myKey)```

**Comment by justinfagnani** [comment]
And for completeness, option 4 is to not use statics as part of the template cache key at all, but skip them when

**Comment by QuackMartins** [comment]
Hey guys, is there any estimate when it's going to be merged?
I tried to install `lit-html#static` but it didn't work

**Comment by LarsDenBakker** [comment]
Would be great to get this in.

**Comment by sghoweri** [comment]
Any updates on this?

**Comment by daKmoR** [comment]
so we currently have a custom function we called `dhtml` which basically does the same thing but as a prestep before lit-html even is involved. Using it we can have dynamic tags even without this change in lit-html.

Usage it is pretty much the same e.g. 
```js
const tag = dynamicStatic('my-el');
render(dhtml`<${tag} someProp=${prop} other static attributes="foo"></${tag}>`, target);

// this will invoke lit-html as if it was written like this
html`<my-el someProp=${prop} other static attributes="foo"></my-el>`
```

as it's probably not as performant as this solution I'm wondering if we should still publish it as it's the only way to have dynamic tags now?

or do we have a timetable on this? I don't wanna wast time preparing, using and later refactoring if this is not too far out.

**Comment by daKmoR** [comment]
I'm afraid I don't have it right now - different computer... I can post it tomorrow
but it's pretty simple (e.g. no cache or so)
```js
// psoido code
const tag = { d: 'my-tag' };

function dhtml(strings, ...values) {
  values.forEach((value) => {
    if (isObjAndHasOnlyDasKey(value)) {
      // add to corresponding strings
      // remove from values
    }
  });
  return html(modifiedStrings, ...modifiedValues);
}
```

**Comment by daKmoR** [comment]
how come? it should be the same as just writing it as normal string?

also not a problem for our usecase as it's only for demos and tests...

anyways is there a timeframe?

so

**Comment by ruphin** [comment]
A naive implementation like that without a cache be very slow. The way tagged template string literals work, is that for each evaluation of the tagged template string, the "tag" function is called with an array of strings as the first parameter. This array object is non-extensible, and it is the exact same object each time the same template is evaluated. Lit-html uses this object equality to determine if it has already processed a certain template string literal or not.

Since the Array is not extensible, it is not possible to "change" the template strings before passing it on to lit-html. You can only create a new Array, which means lit-html considers every evaluation of that template as a new one, and has to process each template again, removing any ability to cache anything.

There is a way to get reasonable performance, but it requires that you pass the exact same strings-array object to lit-html on subsequent evaluations. What you could do is give the `dhtml` function an internal cache that holds a memory of strings-array objects that it generated for each input, and pass the same output to lit-html whenever it gets the same input. You'd have to be very careful in constructing these objects to follow the exact specification outlined in the tagged template string literal spec.

**Comment by justinfagnani** [comment]
Closing in favor of the static-html module in lit-next

**Pull request closed without merging**